source/simulator/solver_schemes.cc

```diff
@@ -564,86 +576,113 @@ namespace aspect
+
+                assemble_stokes_system();
+
+                test_velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                test_pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                test_residual = std::sqrt(test_velocity_residual * test_velocity_residual
+                                          + test_pressure_residual * test_pressure_residual);
+
+                // Determine if the decrease is sufficient.
+                if (test_residual < (1.0 - alpha * lambda) * residual
+                    ||
+                    line_search_iteration >= parameters.max_newton_line_search_iterations
+                    ||
+                    use_picard)
+                  {
+                    pcout << "      Relative nonlinear residual (total Newton system) after nonlinear iteration " << nonlinear_iteration+1
+                          << ": " << test_residual/initial_residual << ", norm of the rhs: " << test_residual
+                          << ", newton_derivative_scaling_factor: " << newton_handler->get_newton_derivative_scaling_factor() << std::endl;
+                    residual = test_residual;
+                    break;
+                  }
+                else
+                  {
+
+                    pcout << "   Line search iteration " << line_search_iteration << ", with norm of the rhs "
+                          << test_residual << " and going to " << (1.0 - alpha * lambda) * residual
+                          << ", relative residual: " << test_residual/initial_residual << std::endl;
+
+                    /**
+                     * The line search step was not sufficient to decrease the residual
+                     * enough, so we take a smaller step to see if it improves the residual.
+                     */
+                    lambda *= (2.0/3.0);// TODO: make a parameter out of this.
+                  }
+
+                line_search_iteration++;
+                Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
+                       ExcMessage ("This tests the while condition. This condition should "
+                                   "actually never be false, because the break statement "
+                                   "above should have caught it."));
               }
-
-            line_search_iteration++;
-            Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
-                   ExcMessage ("This tests the while condition. This condition should "
-                               "actually never be false, because the break statement "
-                               "above should have caught it."));
+            while (line_search_iteration <= parameters.max_newton_line_search_iterations);
+            // The while condition should actually never be false, because the break statement above should have caught it.
```

Comment:
then change it to ``while(true)`` and remove this comment

Reply:
hmm, this was a discussion in #1707 (@gassmoeller review of 16 May) , @gassmoeller and @bangerth where in favour of making this explicit. I am personally fine with changing it back, if that is the consensus.

Reply:
if that is what they prefer, then leave it as is.